src/shogun/base/class_list.cpp

```diff
@@ -32,39 +32,83 @@
 
 #include <shogun/lib/common.h>
 #include <shogun/base/class_list.h>
-#include <string.h>
+#include <shogun/base/library.h>
+#include <shogun/base/SGObject.h>
 
-REPLACE includes THIS
+#include <mutex>
 
 using namespace shogun;
 using namespace shogun::io;
 
-REPLACE definitions THIS
+using CreateFunction = std::function<std::shared_ptr<SGObject>(EPrimitiveType generic)>;
+using ClassList = std::unordered_map<std::string, CreateFunction>;
 
-REPLACE template_definitions THIS
+static const ClassList create_class_list() {
+	std::mutex mu;
+	ClassList class_list;
 
-REPLACE complex_template_definitions THIS
+	std::lock_guard<std::mutex> lock(mu);
+	for (const auto& plugin: env()->plugins())
+	{
+		try
+		{
+			auto library = load_library(plugin);
+			try
+			{
+				for (const auto& class_name: library.manifest().class_list())
+				{
+					if (class_list.find(class_name) != class_list.end())
+					{
+						io::warn("Not registering {} class of {} as it has been already registered!",
+							class_name, plugin);
+						continue;
+					}
 
-typedef SGObject* (*CreateFunction)(EPrimitiveType generic);
+					SG_TRACE("adding {} class of {} to class list", class_name, plugin);
+					class_list.emplace(class_name, [plugin, class_name](EPrimitiveType generic) {
+						// TODO: add library handle caching
+						auto lib = load_library(plugin);
+						auto metaclass = lib.manifest().class_by_name<SGObject>(class_name);
```

Comment:
@lisitsyn ok so if we wanna support(?) old school `SGO create("classname")` then things need to be exported as SGObject based... or i guess at least have in the manifest something like:
```
        template <typename T, enable_if_t<is_base_v<SGObject, T>>* = nullptr>
        MetaClass<SGObject> class_by_name(const std::string& name) const
        {
            auto clazz = find_class(name);
            return clazz.cast<MetaClass<SGObject>>();
        }
```